src/NHibernate/Multi/ICachingInformation.cs

```diff
@@ -68,4 +74,21 @@ public interface ICachingInformation
 		/// <param name="cacheBatcher">A cache batcher.</param>
 		void SetCacheBatcher(CacheBatcher cacheBatcher);
 	}
+
+	internal static class CachingInformationExtensions
+	{
+		// 6.0 TODO: Move to ICachingInformation as a Property
+		public static IType[] GetCacheTypes(this ICachingInformation cachingInformation)
+		{
+			var loaderProperty = cachingInformation.GetType().GetProperty("Loader");
+			if (loaderProperty?.GetValue(cachingInformation) is Loader.Loader loader)
```

Comment:
Avoiding the reflection performance hit is surely worth switching `protected class QueryInfo` to `protected internal` in order to do a cast here instead.

Reply:
Unfortunately it is not possible as `QueryInfo` is a nested class of a generic type and the generic argument would be needed in order to get rid of the reflection but the callers are not able to provide it.

Reply:
Well as an option you can introduce temp interface. Somethinkg like:
// 6.0 TODO: merge into 'ICachingInformation'.
public interface ICachingInformationWithPropertyFetch
{
	IType[] GetCacheTypes();
}

Reply:
Added the proposed interface.